src/metabase/api/dashboard.clj

```diff
@@ -580,37 +646,65 @@
                                                                [:target       :any]]]]]
    [:series             {:optional true} [:maybe [:sequential map?]]]])
 
+(def ^:private UpdatedDashboardTab
+  [:map
+   ;; id can be negative, it indicates a new card and BE should create them
+   [:id   ms/Int]
+   [:name ms/NonBlankString]])
+
+(defn- update-dashboard-with-tabs!
+  [dashboard new-tabs new-cards]
+  (let [{:keys [temp->real-tab-ids
+                deleted-tab-ids]}  (do-update-tabs! dashboard (:ordered_tabs dashboard) new-tabs)
+        current-cards              (if (seq deleted-tab-ids)
+                                       (remove (fn [card]
+                                                 (contains? deleted-tab-ids (:dashboard_tab_id card)))
+                                               (:ordered_cards dashboard))
+                                       (:ordered_cards dashboard))
```

Comment:
when deleting a tab, all of its dash cards will be deleted on cascade so there is no need to re-delete them.

I don't think it's gonna fail but I think it's sensible to provide `do-update-dashcards` with the correct set of `current-cards`

Reply:
We can assume that there are no dashcards in `dashboard` that are in a deleted tab. So you can remove this expression.

Reply:
this is removed from `current-cards`, not the `cards` sent by FE so this is needed. another way is we could hydrate current cards after deletion but I don't think that's necessary.

Reply:
Sorry, I understand now. So this is because you're deleting the tabs (which delete cards) before step where the cards are deleted. How about inlining `do-update-tabs!` and deleting the tabs after the cards like this? Removes 15ish LOC and makes this remove redundant
```
(defn- update-dashboard-with-tabs!
  [dashboard new-tabs new-cards]
  (when-not (= (range (count new-tabs))
               (sort (map :position new-tabs)))
    (throw (ex-info (tru "Tab positions must be sequential and start at 0")
                    {:status-code 400})))
  (let [{:keys [to-create to-update to-delete]} (classify-changes (:ordered_tabs dashboard) new-tabs)
        temp-tab-id->tab-id                     (when (seq to-create)
                                                  (create-tabs! dashboard to-create))
        _                                       (when (seq to-update)
                                                  (update-tabs! dashboard to-update))
        new-cards                 (cond->> new-cards
                                    ;; fixup the temporary tab ids with the real ones
                                    (seq temp-tab-id->tab-id)
                                    (map (fn [card]
                                           (if-let [real-tab-id (get temp-tab-id->tab-id (:dashboard_tab_id card))]
                                             (assoc card :dashboard_tab_id real-tab-id)
                                             card))))]
    (do-update-dashcards! dashboard (:ordered_cards dashboard) new-cards)
    (when-let [to-delete-ids (seq (map :id to-delete))]
      (delete-tabs! to-delete-ids))))
```

Reply:
I don't think it's more readable tbh, this version mixes the updates on `tabs->cards->tabs`. I think updating `tabs->cards` make the flow less complicated and feel more natural since `cards belonged to tabs`.

Reply:
Fair enough!